File: app/src/main/java/com/tutv/android/repository/UserRepository.java
package com.tutv.android.repository;

import android.content.SharedPreferences;
import android.util.Base64;

import com.tutv.android.datasource.retrofit.endpoint.UserAPI;
import com.tutv.android.db.dao.UserDao;
import com.tutv.android.domain.Episode;
import com.tutv.android.domain.Season;
import com.tutv.android.domain.User;

import io.reactivex.Single;

public class UserRepository {
    private final UserDao userDao;
    private final UserAPI userAPI;
    private final SharedPreferences authenticationSharedPreferences;

    public UserRepository(UserDao userDao, UserAPI userAPI, SharedPreferences authenticationSharedPreferences) {
        this.userDao = userDao;
        this.userAPI = userAPI;
        this.authenticationSharedPreferences = authenticationSharedPreferences;
    }

    public Single<User> login(String mail, String password) {
        String credentials = mail + ":" + password;
        String encodedCredentials = android.util.Base64.encodeToString(credentials.getBytes(), Base64.NO_WRAP);

        return userAPI.getCurrentUser("Basic " + encodedCredentials);
    }

    public Single<User> getCurrentUser() {
        return getCurrentUser(true);
    }

    public Single<User> getCurrentUser(boolean invokeLoginScreenOnFail) {
        if(invokeLoginScreenOnFail)
            return userAPI.getCurrentUser();
        else
            return userAPI.getCurrentUserNoForceLoginScreenOnFail();
    }

    public void logout() {
        SharedPreferences.Editor editor = authenticationSharedPreferences.edit();
        editor.remove("Token");
        editor.commit();
    }
}

File: app/src/main/java/com/tutv/android/repository/UserRepository.kt
package com.tutv.android.repository

import com.tutv.android.db.dao.UserDao
import com.tutv.android.datasource.retrofit.endpoint.UserAPI
import android.content.SharedPreferences
import android.util.Base64
import com.tutv.android.domain.User
import io.reactivex.Single

class UserRepository(
    private val userDao: UserDao,
    private val userAPI: UserAPI,
    private val authenticationSharedPreferences: SharedPreferences
) {
    fun login(mail: String, password: String): Single<User> {
        val encodedCredentials = with("$mail:$password") { Base64.encodeToString(this.toByteArray(), Base64.NO_WRAP) }
        return userAPI.getCurrentUser("Basic $encodedCredentials")
    }

    val currentUser: Single<User>
        get() = getCurrentUser(true)

    fun getCurrentUser(invokeLoginScreenOnFail: Boolean): Single<User> {
        return if (invokeLoginScreenOnFail) userAPI.currentUser else userAPI.currentUserNoForceLoginScreenOnFail
    }

    fun logout() {
        authenticationSharedPreferences.edit().apply {
            remove("Token")
            commit()
        }
    }
}